### PR TITLE
jibri: add autoscaler-sidecar service support

### PR DESCRIPTION
**Pull request by aaronkvanmeerten**

### CODE REVIEW

File: jibri.yml [hunk 1]
```diff
@@ -13,6 +13,15 @@ services:
             - CHROMIUM_FLAGS
             - DISPLAY=:0
             - ENABLE_STATS_D
+            - AUTOSCALER_URL
```
Comment (saghul) [inline jibri.yml:16]: Alpha sort please.

File: jibri/rootfs/defaults/autoscaler-sidecar.config [hunk 2]
```diff
@@ -0,0 +1,14 @@
+export PORT={{ .Env.AUTOSCALER_SIDECAR_PORT }}
```
Comment (saghul) [inline jibri/rootfs/defaults/autoscaler-sidecar.config:1]: This port is not present in the compose ports / expose section. Does it need to? Also no default?

File: jibri/rootfs/defaults/autoscaler-sidecar.config [hunk 3]
```diff
@@ -0,0 +1,14 @@
...
+export GRACEFUL_SHUTDOWN_SCRIPT="/opt/jitsi/jibri/wait_graceful_shutdown.sh"
+export TERMINATE_SCRIPT="/opt/jitsi/jibri/shutdown.sh"
+export RECONFIGURE_SCRIPT=
```
Comment (saghul) [inline jibri/rootfs/defaults/autoscaler-sidecar.config:4]: ?

File: jibri/rootfs/defaults/autoscaler-sidecar.config [hunk 4]
```diff
@@ -0,0 +1,14 @@
...
+export ENABLE_REPORT_STATS=true
+export POLLING_URL="{{ .Env.AUTOSCALER_URL  }}/sidecar/poll"
+export STATUS_URL="{{ .Env.AUTOSCALER_URL }}/sidecar/status"
+export STATS_RETRIEVE_URL="http://localhost:2222/jibri/api/v1.0/health"
```
Comment (saghul) [inline jibri/rootfs/defaults/autoscaler-sidecar.config:8]: Isn't that port conficurable?

File: jibri/rootfs/defaults/autoscaler-sidecar.config [hunk 5]
```diff
@@ -0,0 +1,14 @@
...
+export STATS_REPORT_URL="{{ .Env.AUTOSCALER_URL }}/sidecar/stats"
+export ASAP_SIGNING_KEY_FILE="{{ .Env.AUTOSCALER_SIDECAR_KEY_FILE }}"
+export ASAP_JWT_KID="{{ .Env.AUTOSCALER_SIDECAR_KEY_ID }}"
+export INSTANCE_TYPE="jibri"
```
Comment (saghul) [inline jibri/rootfs/defaults/autoscaler-sidecar.config:12]: Should this be configurable?

File: jibri.yml [hunk 6]
```diff
@@ -13,6 +13,15 @@ services:
...
+            - AUTOSCALER_SIDECAR_KEY_FILE
+            - AUTOSCALER_SIDECAR_KEY_ID
+            - AUTOSCALER_SIDECAR_PORT
+            - AUTOSCALER_SIDECAR_INSTANCE_ID
+            - AUTOSCALER_SIDECAR_REGION
+            - AUTOSCALER_SIDECAR_GROUP_NAME
+            - LOCAL_ADDRESS
```
Comment (saghul) [inline jibri.yml:23]: What is this supposed to do?

File: jibri/rootfs/defaults/autoscaler-sidecar.config [hunk 7]
```diff
@@ -0,0 +1,14 @@
...
+export INSTANCE_ID="{{ .Env.AUTOSCALER_SIDECAR_INSTANCE_ID }}"
+export INSTANCE_METADATA='{"environment":"{{ .Env.XMPP_ENV_NAME }}","region":"{{ .Env.AUTOSCALER_SIDECAR_REGION }}","group":"{{ .Env.AUTOSCALER_SIDECAR_GROUP_NAME }}","name":"{{ .Env.JIBRI_INSTANCE_ID }}","version":"{{ .Env.JIBRI_VERSION }}","privateIp":"{{ .Env.LOCAL_ADDRESS }}"}'
```
Comment (saghul) [inline jibri/rootfs/defaults/autoscaler-sidecar.config:14]: is the private IP relevant?

File: jibri/rootfs/etc/cont-init.d/10-config [hunk 8]
```diff
@@ -30,6 +30,40 @@ fi
...
+    fi
+
+    if [ -f "$AUTOSCALER_SIDECAR_KEY_FILE" ]; then
+        echo "AUTOSCALER_URL found, enabling autoscaler sidecar"
+
+        export JIBRI_VERSION="$(dpkg -s jibri | grep Version | awk '{print $2}' | sed 's/..$//')"
+
+        [ -z "$AUTOSCALER_SIDECAR_PORT" ] && export AUTOSCALER_SIDECAR_PORT="6000"
+        [ -z "$JIBRI_WEBHOOK_SUBSCRIBERS" ] && export JIBRI_WEBHOOK_SUBSCRIBERS="http://localhost:$AUTOSCALER_SIDECAR_PORT/hook"
+        [ -z "$AUTOSCALER_SIDECAR_INSTANCE_ID" ] && export AUTOSCALER_SIDECAR_INSTANCE_ID="$JIBRI_INSTANCE_ID"
```
Comment (saghul) [inline jibri/rootfs/etc/cont-init.d/10-config:50]: Does it make sense for the 2 ids to be different?

File: jibri/rootfs/defaults/autoscaler-sidecar.config (hunk 2 above)
Comment (aaronkvanmeerten) [reply jibri/rootfs/defaults/autoscaler-sidecar.config:1]: It's only used internally by the jibri running in the same container.  It's default is set in the shell script

File: jibri/rootfs/defaults/autoscaler-sidecar.config (hunk 5 above)
Comment (aaronkvanmeerten) [reply jibri/rootfs/defaults/autoscaler-sidecar.config:12]: No this is the type of instance we are running for the jibri container.  For other containers we will set it differently.

File: jibri/rootfs/defaults/autoscaler-sidecar.config (hunk 7 above)
Comment (aaronkvanmeerten) [reply jibri/rootfs/defaults/autoscaler-sidecar.config:14]: Only for informational purposes to identify the container or node, can be filled in by nomad, kube etc as desired.

File: jibri/rootfs/defaults/autoscaler-sidecar.config (hunk 4 above)
Comment (aaronkvanmeerten) [reply jibri/rootfs/defaults/autoscaler-sidecar.config:8]: Why yes, it is!

File: jibri.yml (hunk 1 above)
Comment (aaronkvanmeerten) [reply jibri.yml]: done

**Review by saghul — APPROVED** [review]

**Comment by jvkassi** [comment]
how does autoscaler works ?

**Comment by aaronkvanmeerten** [comment]
> how does autoscaler works ?

It's a work in progress, but the basic idea is that the sidecar sends stats to a microservice which then controls the number of instances and their various states (drain vs. in use).  The existing microservice currently only supports nomad deployments of containers, but could probably be extended to do kubernetes as well.  The source is here: https://github.com/jitsi/jitsi-autoscaler

**Pull request merged**